### PR TITLE
OSSMDOC-395: Configure Federation for failover.

### PR DESCRIPTION
**Pull request by JStickler**
Add documentation for how to configure Federation for failover.
Cleaned up some leftover file notes that should have been removed before release.

Preview is here ->  https://deploy-preview-41725--osdocs.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-federation#ossm-federation-config-failover_federation

Eng review - rcernich
QE review - yxun
Peer review - sagidlow

### CODE REVIEW

**Comment by rcernich** [comment]
This content looks correct, but I'm not sure if we've verified this through any testing (dev or qe).

File: modules/ossm-federation-config-failover.adoc [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  trafficPolicy:
+    loadBalancer:
+      localityLbSetting:
+        enabled: true
+        failover:
+          - from: us-east
+          to: us-west
+    outlierDetection:
+      consecutive5xxErrors: 1
+      interval: 1s
```
[the remark this answers is not in the document]
Comment (JStickler) [reply modules/ossm-federation-config-failover.adoc]: Updated the value.

**Comment by bergerhoffer** [comment]
The `branch/enterprise-4.11` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.10. And any PR going into `main` must also target the latest version branch (enterprise-4.11).

If the update in your PR does NOT apply to version 4.11 onward, please retarget this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

File: modules/ossm-federation-config-failover.adoc [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  namespace: green-mesh-system
+spec:
+  host: "*.green-mesh-system.svc.cluster.local"
+  trafficPolicy:
+    loadBalancer:
+      localityLbSetting:
+        enabled: true
+        failover:
+          - from: us-east
+          to: us-west
```
Comment (yxun) [inline modules/ossm-federation-config-failover.adoc:161]: @JStickler There is a syntax indentation error in line 161.
The `to` should align with `from` . Need to add two more spaces.
Reference:
https://istio.io/latest/docs/tasks/traffic-management/locality-load-balancing/failover/#configure-locality-failover

File: modules/ossm-federation-config-failover.adoc [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+[id="ossm-federation-config-failover_{context}"]
+= Configuring a federated mesh for failover
+
+Failover is the ability to switch automatically and seamlessly to a reliable backup system, for example another server. In the case of a federated mesh, you can configure a service in one mesh to failover to a service in another mesh.
+
+You configure Federation for failover by setting the `importAsLocal` and `locality` settings in a `ServiceImportSet` resource and then configuring a `DestinationRule` that configures failover for the service to the locality specified in the `ServiceImportSet`.
+
+.Prerequisites
+
+* Two or more OpenShift Container Platform 4.6 or above clusters already networked and federated.
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:15]:
```suggestion
* Two or more {product-title} 4.6 or above clusters already networked and federated.
```

File: modules/ossm-federation-config-failover.adoc [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      alias:
+        # service will be imported as ratings.bookinfo.svc.red-mesh-imports.local
+        namespace: bookinfo
+        name: ratings
+  #Locality within which imported services should be associated.
+  locality:
+    region: us-west
+----
+
+.Table ImportedServiceLocality Fields
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:51]:
```suggestion
.`ImportedServiceLocality` fields table
```
Add an introduction sentence here to introduce this table.

File: modules/ossm-federation-config-failover.adoc [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+|zone:
+|Zone within which imported services are located.  If Zone is specified, Region must also be specified.
+|string
+|===
+
+
+.Procedure
+
+. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command.
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:71]:
```suggestion
. Log in to the {product-title} CLI as a user with the `cluster-admin` role, enter the following command:
```

> https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

File: modules/ossm-federation-config-failover.adoc [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+You configure Federation for failover by setting the `importAsLocal` and `locality` settings in a `ServiceImportSet` resource and then configuring a `DestinationRule` that configures failover for the service to the locality specified in the `ServiceImportSet`.
+
+.Prerequisites
+
+* Two or more OpenShift Container Platform 4.6 or above clusters already networked and federated.
+* `ExportServiceSet` resources already created for each mesh peer in the federated mesh.
+* `ImportServiceSet` resources already created for each mesh peer in the federated mesh.
+* An account with the cluster-admin role.
+
+== Configuring an ImportServiceSet for failover
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:20]: I would put `==Configuring an ImportServiceSet for failover `into a new module and use that module as a procedure module. You cannot have multiple headings in a procedure file.

File: modules/ossm-federation-config-failover.adoc [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+----
++
+For example, `green-mesh-system`.
++
+[source,terminal]
+----
+$ oc project green-mesh-system
+----
++
+. Run the following command to edit the `ImportServiceSet` file, where `<ImportServiceSet.yaml>` includes a full path to the file you want to edit:
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:92]: To maintain some consistency with the above procedure maybe rework this to include ",enter the following command"? Something like this?

File: modules/ossm-federation-config-failover.adoc [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+$ oc edit -n green-mesh-system -f import-from-red-mesh.yaml
+----
+. Modify the file:
+.. Set `spec.importRules.importAsLocal` to `true`.
+.. Set `spec.locality` to a `region`, `zone`, or `subzone`.
+.. Save your changes.
+
+//Anything else we need to do to deploy/update the changed file?
+
+== Configuring a DestinationRule for failover
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:112]: I would put `==Configuring a DestinationRule for failover` into a new module and use that module as a procedure module.  You cannot have multiple headings in a procedure file.

File: modules/ossm-federation-config-failover.adoc [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        failover:
+          - from: us-east
+            to: us-west
+    outlierDetection:
+      consecutive5xxErrors: 1
+      interval: 5m
+      baseEjectionTime: 1m
+----
+
+. Run the following command to deploy the `DestinationRule`, where `<DestinationRule>` includes the full path to your file.
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:168]:
```suggestion
. Deploy the `DestinationRule`, where `<DestinationRule>` includes the full path to your file, enter the following command:
```

File: modules/ossm-federation-config-failover.adoc (hunk 7 above)
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:92]:
```suggestion
. Edit the `ImportServiceSet` file, where `<ImportServiceSet.yaml>` includes a full path to the file you want to edit, enter the following command:
```

File: modules/ossm-federation-config-failover.adoc [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+.Procedure
+
+. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command.
++
+[source,terminal]
+----
+$ oc login --username=<NAMEOFUSER> <API token> https://<HOSTNAME>:6443
+----
++
+. Change to the project where you installed the control plane.
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:78]:
```suggestion
. Change to the project where you installed the control plane, enter the following command:
```

File: modules/ossm-federation-config-failover.adoc [hunk 11]
```diff
@@ -0,0 +1,180 @@
+////
+This module included in the following assemblies:
+* service_mesh/v2x/ossm-federation.adoc
+////
+:_content-type: PROCEDURE ????????????????
+[id="ossm-federation-config-failover_{context}"]
+= Configuring a federated mesh for failover
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover.adoc:7]:
```suggestion
= Configure a federated mesh for failover
```
To use in a concept file I would drop the "ing".

**Review by sagidlow — COMMENTED** [review]
Let me know if you have any questions :)

File: modules/ossm-federation-config-failover.adoc [hunk 12]
```diff
@@ -0,0 +1,180 @@
+////
+This module included in the following assemblies:
+* service_mesh/v2x/ossm-federation.adoc
+////
+:_content-type: PROCEDURE ????????????????
```
[the remark this answers is not in the document]
Comment (JStickler) [reply modules/ossm-federation-config-failover.adoc]: Done

File: modules/ossm-federation-config-failover-overview.adoc [hunk 13]
```diff
@@ -0,0 +1,18 @@
+////
+This module included in the following assemblies:
+* service_mesh/v2x/ossm-federation.adoc
+////
+:_content-type: CONCEPT
+[id="ossm-federation-config-failover-overview_{context}"]
+= Configuring a federated mesh for failover
```
Comment (sagidlow) [inline modules/ossm-federation-config-failover-overview.adoc:7]: Configure?

File: modules/ossm-federation-config-destinationrule-failover.adoc [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+Create a `DestinationRule` resource that configures the following:
+
+* Outlier detection for the service. This is required in order for failover to function properly. In particular, it configures the sidecar proxies to know when endpoints for a service are unhealthy, eventually triggering a failover to the next locality.
+
+* Failover policy between regions. This ensures that failover beyond a region boundary will behave predictably.
+
+.Procedure
+
+. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command. Then, enter your username and password when prompted.
```
Comment (sagidlow) [inline modules/ossm-federation-config-destinationrule-failover.adoc:17]:
```suggestion
. Log in to the {product-title} CLI with `cluster-admin` permissions. Enter your username and password when prompted.
```
I tried to simplify this a little. Definitely not required.

**Review by sagidlow — COMMENTED** [review]
Just a couple more nit comments. :) Otherwise, lgtm.

File: modules/ossm-federation-config-importserviceset-failover.adoc [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        # service will be imported as ratings.bookinfo.svc.red-mesh-imports.local
+        namespace: bookinfo
+        name: ratings
+  #Locality within which imported services should be associated.
+  locality:
+    region: us-west
+----
+
+.`ImportedServiceLocality` fields table
+|===
```
Comment (sagidlow) [inline modules/ossm-federation-config-importserviceset-failover.adoc:39]: I would still add an introduction to this table.

**Pull request merged**